File: app/src/main/java/com/example/data/repository/RestaurantRepository.kt
package com.example.data.repository

import com.example.data.database.AppDatabase
import com.example.data.model.CartItem
import com.example.data.model.CategoryEntity
import com.example.data.model.MenuEntity
import com.example.data.model.OrderEntity
import com.example.data.model.OrderItemEntity
import com.example.data.model.OrderStatus
import com.example.data.model.OrderType
import com.example.data.model.OrderWithItems
import com.example.data.model.PaymentMethod
import com.example.data.model.TableEntity
import com.example.data.model.TableStatus
import kotlinx.coroutines.flow.Flow
import java.text.NumberFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlin.random.Random

class RestaurantRepository(private val database: AppDatabase) {

    private val categoryDao = database.categoryDao()
    private val menuDao = database.menuDao()
    private val tableDao = database.tableDao()
    private val orderDao = database.orderDao()

    val allCategories: Flow<List<CategoryEntity>> = categoryDao.getAllCategories()
    val allMenus: Flow<List<MenuEntity>> = menuDao.getAllMenus()
    val allTables: Flow<List<TableEntity>> = tableDao.getAllTables()
    val allOrdersWithItems: Flow<List<OrderWithItems>> = orderDao.getAllOrdersWithItems()

    fun getMenusByCategory(categoryId: Int): Flow<List<MenuEntity>> = menuDao.getMenusByCategory(categoryId)
    fun searchMenus(query: String): Flow<List<MenuEntity>> = menuDao.searchMenus(query)
    fun getOrdersByStatus(status: OrderStatus): Flow<List<OrderWithItems>> = orderDao.getOrdersByStatus(status)

    suspend fun getOrderById(orderId: Long): OrderWithItems? = orderDao.getOrderWithItemsById(orderId)

    // === Menu Management ===
    suspend fun insertMenu(menu: MenuEntity): Long = menuDao.insertMenu(menu)
    suspend fun updateMenu(menu: MenuEntity) = menuDao.updateMenu(menu)
    suspend fun deleteMenu(menu: MenuEntity) = menuDao.deleteMenu(menu)
    suspend fun updateStock(menuId: Int, newStock: Int) {
        val menu = menuDao.getMenuById(menuId)
        if (menu != null) {
            menuDao.updateMenu(menu.copy(stock = newStock))
        }
    }

    // === Category Management ===
    suspend fun insertCategory(category: CategoryEntity): Long = categoryDao.insertCategory(category)
    suspend fun deleteCategory(category: CategoryEntity) = categoryDao.deleteCategory(category)

    // === Table Management ===
    suspend fun insertTable(table: TableEntity): Long = tableDao.insertTable(table)
    suspend fun updateTable(table: TableEntity) = tableDao.updateTable(table)
    suspend fun updateTableStatus(tableId: Int, status: TableStatus, activeOrderId: Long? = null) {
        tableDao.updateTableStatus(tableId, status, activeOrderId)
    }

    // === Checkout & Payment Engine (100% Offline) ===
    suspend fun processCheckout(
        cartItems: List<CartItem>,
        table: TableEntity?,
        customerName: String,
        orderType: OrderType,
        discountPercent: Double,
        taxPercent: Double,
        servicePercent: Double,
        paymentMethod: PaymentMethod,
        amountPaid: Double,
        notes: String = "",
        cashierName: String = "Kasir Resto",
        isPaidNow: Boolean = true
    ): OrderWithItems {
        val timestamp = System.currentTimeMillis()
        val dateStr = SimpleDateFormat("yyyyMMdd", Locale.getDefault()).format(Date(timestamp))
        val randomSuffix = Random.nextInt(1000, 9999)
        val invoiceNumber = "INV-$dateStr-$randomSuffix"

        val subtotal = cartItems.sumOf { it.subtotal }
        val discountAmount = (subtotal * (discountPercent / 100.0))
        val afterDiscount = subtotal - discountAmount
        val taxAmount = (afterDiscount * (taxPercent / 100.0))
        val serviceAmount = (afterDiscount * (servicePercent / 100.0))
        val grandTotal = afterDiscount + taxAmount + serviceAmount

        val finalStatus = if (isPaidNow) OrderStatus.PAID else OrderStatus.PENDING
        val finalPaid = if (isPaidNow) amountPaid else 0.0
        val changeAmount = if (isPaidNow && amountPaid >= grandTotal) amountPaid - grandTotal else 0.0

        val referenceNumber = when (paymentMethod) {
            PaymentMethod.QRIS_OFFLINE -> "QRIS-OFF-$timestamp"
            PaymentMethod.DEBIT_CARD -> "EDC-DEBIT-${Random.nextInt(100000, 999999)}"
            PaymentMethod.CREDIT_CARD -> "EDC-CC-${Random.nextInt(100000, 999999)}"
            PaymentMethod.TRANSFER_BANK -> "TF-VA-${Random.nextInt(10000000, 99999999)}"
            PaymentMethod.CASH -> "CASH-$randomSuffix"
        }

        val order = OrderEntity(
            invoiceNumber = invoiceNumber,
            tableId = table?.id,
            tableName = table?.let { "Meja ${it.tableNumber}" } ?: (if (orderType == OrderType.TAKEAWAY) "Takeaway / Bungkus" else "Delivery"),
            customerName = if (customerName.isBlank()) (table?.let { "Tamu Meja ${it.tableNumber}" } ?: "Pelanggan") else customerName,
            orderType = orderType,
            subtotal = subtotal,
            discountPercent = discountPercent,
            discountAmount = discountAmount,
            taxRatePercent = taxPercent,
            taxAmount = taxAmount,
            serviceRatePercent = servicePercent,
            serviceAmount = serviceAmount,
            grandTotal = grandTotal,
            paymentMethod = paymentMethod,
            amountPaid = finalPaid,
            changeAmount = changeAmount,
            status = finalStatus,
            notes = notes,
            cashierName = cashierName,
            referenceNumber = referenceNumber,
            createdAt = timestamp
        )

        val orderId = orderDao.insertOrder(order)

        // Insert Order Items and reduce stock atomically
        val orderItems = cartItems.map { item ->
            menuDao.reduceStock(item.menu.id, item.quantity)
            OrderItemEntity(
                orderId = orderId,
                menuId = item.menu.id,
                menuName = item.menu.name,
                categoryName = item.menu.categoryName,
                unitPrice = item.menu.price,
                quantity = item.quantity,
                subtotal = item.subtotal,
                notes = item.notes
            )
        }
        orderDao.insertOrderItems(orderItems)

        // Update table status if dine in
        table?.let {
            if (isPaidNow) {
                // If paid immediately, table remains available
                tableDao.updateTableStatus(it.id, TableStatus.AVAILABLE, null)
            } else {
                // If saved pending, mark table as occupied with active order
                tableDao.updateTableStatus(it.id, TableStatus.OCCUPIED, orderId)
            }
        }

        return OrderWithItems(order = order.copy(id = orderId), items = orderItems)
    }

    // === Settle Pending Bill ===
    suspend fun settlePendingOrder(
        orderId: Long,
        paymentMethod: PaymentMethod,
        amountPaid: Double
    ): OrderWithItems? {
        val orderWithItems = orderDao.getOrderWithItemsById(orderId) ?: return null
        val order = orderWithItems.order
        val grandTotal = order.grandTotal
        val change = if (amountPaid >= grandTotal) amountPaid - grandTotal else 0.0

        val referenceNumber = when (paymentMethod) {
            PaymentMethod.QRIS_OFFLINE -> "QRIS-OFF-${System.currentTimeMillis()}"
            PaymentMethod.DEBIT_CARD -> "EDC-DEBIT-${Random.nextInt(100000, 999999)}"
            PaymentMethod.CREDIT_CARD -> "EDC-CC-${Random.nextInt(100000, 999999)}"
            PaymentMethod.TRANSFER_BANK -> "TF-VA-${Random.nextInt(10000000, 99999999)}"
            PaymentMethod.CASH -> "CASH-${Random.nextInt(1000, 9999)}"
        }

        val updatedOrder = order.copy(
            status = OrderStatus.PAID,
            paymentMethod = paymentMethod,
            amountPaid = amountPaid,
            changeAmount = change,
            referenceNumber = referenceNumber
        )
        orderDao.updateOrder(updatedOrder)

        // Release table if occupied
        order.tableId?.let { tableId ->
            tableDao.updateTableStatus(tableId, TableStatus.AVAILABLE, null)
        }

        return OrderWithItems(order = updatedOrder, items = orderWithItems.items)
    }

    // === Void / Cancel Order ===
    suspend fun cancelOrder(orderId: Long) {
        val orderWithItems = orderDao.getOrderWithItemsById(orderId) ?: return
        // Restore stock
        for (item in orderWithItems.items) {
            menuDao.restoreStock(item.menuId, item.quantity)
        }
        orderDao.cancelOrder(orderId)

        // Free table if associated
        orderWithItems.order.tableId?.let { tableId ->
            tableDao.updateTableStatus(tableId, TableStatus.AVAILABLE, null)
        }
    }

    companion object {
        fun formatRupiah(amount: Double): String {
            val format = NumberFormat.getCurrencyInstance(Locale("id", "ID"))
            return format.format(amount).replace("Rp", "Rp ").replace(",00", "")
        }

        fun formatTimestamp(timestamp: Long): String {
            val sdf = SimpleDateFormat("dd MMM yyyy, HH:mm", Locale("id", "ID"))
            return sdf.format(Date(timestamp))
        }

        fun formatDateOnly(timestamp: Long): String {
            val sdf = SimpleDateFormat("EEEE, dd MMMM yyyy", Locale("id", "ID"))
            return sdf.format(Date(timestamp))
        }
    }
}

File: lib/repositories/restaurant_repository.dart
import 'dart:math';
import 'package:intl/intl.dart';
import '../database/app_database.dart';
import '../models/cart_item_model.dart';
import '../models/category_model.dart';
import '../models/menu_model.dart';
import '../models/order_model.dart';
import '../models/table_model.dart';

class RestaurantRepository {
  final AppDatabase database;

  RestaurantRepository({AppDatabase? db}) : database = db ?? AppDatabase.instance;

  // === Fetch Operations ===
  Future<List<CategoryModel>> getAllCategories() => database.getAllCategories();
  Future<List<MenuModel>> getAllMenus() => database.getAllMenus();
  Future<List<TableModel>> getAllTables() => database.getAllTables();
  Future<List<OrderWithItemsModel>> getAllOrdersWithItems() => database.getAllOrdersWithItems();
  Future<OrderWithItemsModel?> getOrderById(int orderId) => database.getOrderWithItemsById(orderId);

  // === Menu Management ===
  Future<int> insertMenu(MenuModel menu) => database.insertMenu(menu);
  Future<int> updateMenu(MenuModel menu) => database.updateMenu(menu);
  Future<int> deleteMenu(int id) => database.deleteMenu(id);
  Future<void> updateStock(int menuId, int newStock) async {
    final menu = await database.getMenuById(menuId);
    if (menu != null) {
      await database.updateMenu(menu.copyWith(stock: newStock));
    }
  }

  // === Category Management ===
  Future<int> insertCategory(CategoryModel category) => database.insertCategory(category);
  Future<int> deleteCategory(int id) => database.deleteCategory(id);

  // === Table Management ===
  Future<int> insertTable(TableModel table) => database.insertTable(table);
  Future<int> updateTable(TableModel table) => database.updateTable(table);
  Future<void> updateTableStatus(int tableId, TableStatus status, [int? activeOrderId]) =>
      database.updateTableStatus(tableId, status, activeOrderId);

  // === Checkout & Payment Engine (100% Offline) ===
  Future<OrderWithItemsModel> processCheckout({
    required List<CartItemModel> cartItems,
    TableModel? table,
    String customerName = 'Pelanggan',
    OrderType orderType = OrderType.dineIn,
    double discountPercent = 0.0,
    double taxPercent = 10.0,
    double servicePercent = 0.0,
    PaymentMethod paymentMethod = PaymentMethod.cash,
    double amountPaid = 0.0,
    String notes = '',
    String cashierName = 'Kasir Resto',
    bool isPaidNow = true,
  }) async {
    final now = DateTime.now();
    final dateStr = DateFormat('yyyyMMdd').format(now);
    final randomSuffix = 1000 + Random().nextInt(9000);
    final invoiceNumber = 'INV-$dateStr-$randomSuffix';

    final subtotal = cartItems.fold<double>(0.0, (sum, item) => sum + item.subtotal);
    final discountAmount = subtotal * (discountPercent / 100.0);
    final afterDiscount = subtotal - discountAmount;
    final taxAmount = afterDiscount * (taxPercent / 100.0);
    final serviceAmount = afterDiscount * (servicePercent / 100.0);
    final grandTotal = afterDiscount + taxAmount + serviceAmount;

    final finalStatus = isPaidNow ? OrderStatus.paid : OrderStatus.pending;
    final finalPaid = isPaidNow ? amountPaid : 0.0;
    final changeAmount = (isPaidNow && amountPaid >= grandTotal) ? amountPaid - grandTotal : 0.0;

    final timestamp = now.millisecondsSinceEpoch;
    String referenceNumber = '';
    switch (paymentMethod) {
      case PaymentMethod.qrisOffline:
        referenceNumber = 'QRIS-OFF-$timestamp';
        break;
      case PaymentMethod.debitCard:
        referenceNumber = 'EDC-DEBIT-${100000 + Random().nextInt(900000)}';
        break;
      case PaymentMethod.creditCard:
        referenceNumber = 'EDC-CC-${100000 + Random().nextInt(900000)}';
        break;
      case PaymentMethod.transferBank:
        referenceNumber = 'TF-VA-${10000000 + Random().nextInt(90000000)}';
        break;
      case PaymentMethod.cash:
        referenceNumber = 'CASH-$randomSuffix';
        break;
    }

    final order = OrderModel(
      invoiceNumber: invoiceNumber,
      tableId: table?.id,
      tableName: table != null
          ? 'Meja ${table.tableNumber}'
          : (orderType == OrderType.takeaway ? 'Takeaway / Bungkus' : 'Delivery'),
      customerName: customerName.trim().isEmpty
          ? (table != null ? 'Tamu Meja ${table.tableNumber}' : 'Pelanggan')
          : customerName.trim(),
      orderType: orderType,
      subtotal: subtotal,
      discountPercent: discountPercent,
      discountAmount: discountAmount,
      taxRatePercent: taxPercent,
      taxAmount: taxAmount,
      serviceRatePercent: servicePercent,
      serviceAmount: serviceAmount,
      grandTotal: grandTotal,
      paymentMethod: paymentMethod,
      amountPaid: finalPaid,
      changeAmount: changeAmount,
      status: finalStatus,
      notes: notes,
      cashierName: cashierName,
      referenceNumber: referenceNumber,
      createdAt: timestamp,
    );

    final orderId = await database.insertOrder(order);

    // Insert Order Items and reduce stock atomically
    final List<OrderItemModel> orderItems = [];
    for (var item in cartItems) {
      if (item.menu.id != null) {
        await database.reduceStock(item.menu.id!, item.quantity);
      }
      orderItems.add(
        OrderItemModel(
          orderId: orderId,
          menuId: item.menu.id ?? 0,
          menuName: item.menu.name,
          categoryName: item.menu.categoryName,
          unitPrice: item.menu.price,
          quantity: item.quantity,
          subtotal: item.subtotal,
          notes: item.notes,
        ),
      );
    }
    await database.insertOrderItems(orderItems);

    // Update table status if dine in
    if (table != null && table.id != null) {
      if (isPaidNow) {
        await database.updateTableStatus(table.id!, TableStatus.available, null);
      } else {
        await database.updateTableStatus(table.id!, TableStatus.occupied, orderId);
      }
    }

    return OrderWithItemsModel(
      order: order.copyWith(id: orderId),
      items: orderItems,
    );
  }

  // === Settle Pending Bill ===
  Future<OrderWithItemsModel?> settlePendingOrder({
    required int orderId,
    required PaymentMethod paymentMethod,
    required double amountPaid,
  }) async {
    final orderWithItems = await database.getOrderWithItemsById(orderId);
    if (orderWithItems == null) return null;

    final order = orderWithItems.order;
    final grandTotal = order.grandTotal;
    final change = (amountPaid >= grandTotal) ? amountPaid - grandTotal : 0.0;

    String referenceNumber = '';
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    switch (paymentMethod) {
      case PaymentMethod.qrisOffline:
        referenceNumber = 'QRIS-OFF-$timestamp';
        break;
      case PaymentMethod.debitCard:
        referenceNumber = 'EDC-DEBIT-${100000 + Random().nextInt(900000)}';
        break;
      case PaymentMethod.creditCard:
        referenceNumber = 'EDC-CC-${100000 + Random().nextInt(900000)}';
        break;
      case PaymentMethod.transferBank:
        referenceNumber = 'TF-VA-${10000000 + Random().nextInt(90000000)}';
        break;
      case PaymentMethod.cash:
        referenceNumber = 'CASH-${1000 + Random().nextInt(9000)}';
        break;
    }

    final updatedOrder = order.copyWith(
      status: OrderStatus.paid,
      paymentMethod: paymentMethod,
      amountPaid: amountPaid,
      changeAmount: change,
      referenceNumber: referenceNumber,
    );
    await database.updateOrder(updatedOrder);

    // Release table if occupied
    if (order.tableId != null) {
      await database.updateTableStatus(order.tableId!, TableStatus.available, null);
    }

    return OrderWithItemsModel(order: updatedOrder, items: orderWithItems.items);
  }

  // === Void / Cancel Order ===
  Future<void> cancelOrder(int orderId) async {
    final orderWithItems = await database.getOrderWithItemsById(orderId);
    if (orderWithItems == null) return;

    // Restore stock
    for (var item in orderWithItems.items) {
      await database.restoreStock(item.menuId, item.quantity);
    }
    await database.cancelOrder(orderId);

    // Free table if associated
    if (orderWithItems.order.tableId != null) {
      await database.updateTableStatus(orderWithItems.order.tableId!, TableStatus.available, null);
    }
  }

  // === Static Helper Formatters ===
  static String formatRupiah(double amount) {
    final formatter = NumberFormat.currency(
      locale: 'id_ID',
      symbol: 'Rp ',
      decimalDigits: 0,
    );
    return formatter.format(amount);
  }

  static String formatTimestamp(int timestamp) {
    final date = DateTime.fromMillisecondsSinceEpoch(timestamp);
    final formatter = DateFormat('dd MMM yyyy, HH:mm', 'id_ID');
    return formatter.format(date);
  }

  static String formatDateOnly(int timestamp) {
    final date = DateTime.fromMillisecondsSinceEpoch(timestamp);
    final formatter = DateFormat('EEEE, dd MMMM yyyy', 'id_ID');
    return formatter.format(date);
  }
}
